Mathlib/Analysis/Meromorphic/Divisor/MeromorphicFunction.lean

```diff
@@ -0,0 +1,124 @@
+      exists_eq_right, Pi.zero_apply, dite_eq_right_iff, WithTop.untopD_eq_self_iff,
+      WithTop.coe_zero]
+    tauto
+
+/-- Definition of the divisor. -/
+theorem divisor_def {f : 𝕜 → E} (hf : MeromorphicOn f U) :
+    divisor f hf z = if hz : z ∈ U then (hf z hz).order.untopD 0 else 0 := rfl
+
+/-- Simplifier lemma: On `U`, the divisor of a function `f` that is meromorphic on `U` evaluates to
+  `order.untopD`. -/
+@[simp]
+lemma divisor_apply {f : 𝕜 → E} (hf : MeromorphicOn f U) (hz : z ∈ U) :
+    divisor f hf z = (hf z hz).order.untopD 0 := by simp_all [hf.divisor_def, hz]
+
+/-!
+## Behavior under Standard Operations
+-/
+
+/-- If orders are finite, the divisor of the scalar product of two meromorphic
+  functions is the sum of the divisors.
+
+  See `MeromorphicOn.exists_order_ne_top_iff_forall` and
+  `MeromorphicOn.order_ne_top_of_isPreconnected` for two convenient criteria to
+  guarantee conditions `h₂f₁` and `h₂f₂`.
+-/
+theorem divisor_smul [CompleteSpace 𝕜] {f₁ : 𝕜 → 𝕜} {f₂ : 𝕜 → E} (h₁f₁ : MeromorphicOn f₁ U)
+    (h₁f₂ : MeromorphicOn f₂ U) (h₂f₁ : ∀ z, (hz : z ∈ U) → (h₁f₁ z hz).order ≠ ⊤)
+    (h₂f₂ : ∀ z, (hz : z ∈ U) → (h₁f₂ z hz).order ≠ ⊤) :
+    divisor (f₁ • f₂) (h₁f₁.smul h₁f₂) = divisor f₁ h₁f₁ + divisor f₂ h₁f₂ := by
+  ext z
+  by_cases hz : z ∈ U
+  · simp_all [(h₁f₁ z hz).order_smul (h₁f₂ z hz)]
+    lift (h₁f₁ z hz).order to ℤ using (h₂f₁ z hz) with a₁ ha₁
+    lift (h₁f₂ z hz).order to ℤ using (h₂f₂ z hz) with a₂ ha₂
+    exact rfl
+  · simp [hz]
+
+/-- If orders are finite, the divisor of the product of two meromorphic
+  functions is the sum of the divisors.
+
+  See `MeromorphicOn.exists_order_ne_top_iff_forall` and
+  `MeromorphicOn.order_ne_top_of_isPreconnected` for two convenient criteria to
+  guarantee conditions `h₂f₁` and `h₂f₂`.
+-/
+theorem divisor_mul [CompleteSpace 𝕜] {f₁ f₂ : 𝕜 → 𝕜} (h₁f₁ : MeromorphicOn f₁ U)
+    (h₁f₂ : MeromorphicOn f₂ U) (h₂f₁ : ∀ z, (hz : z ∈ U) → (h₁f₁ z hz).order ≠ ⊤)
+    (h₂f₂ : ∀ z, (hz : z ∈ U) → (h₁f₂ z hz).order ≠ ⊤) :
+    divisor (f₁ * f₂) (h₁f₁.mul h₁f₂) = divisor f₁ h₁f₁ + divisor f₂ h₁f₂ := by
+  ext z
+  by_cases hz : z ∈ U
+  · simp_all [(h₁f₁ z hz).order_mul (h₁f₂ z hz)]
+    lift (h₁f₁ z hz).order to ℤ using (h₂f₁ z hz) with a₁ ha₁
+    lift (h₁f₂ z hz).order to ℤ using (h₂f₂ z hz) with a₂ ha₂
+    exact rfl
+  · simp [hz]
+
+/-- The divisor of the inverse is the negative of the divisor. -/
+theorem divisor_inv [CompleteSpace 𝕜] {f: 𝕜 → 𝕜} (hf : MeromorphicOn f U) :
+    divisor f⁻¹ hf.inv = -divisor f hf := by
+  ext z
+  by_cases hz : z ∈ U
+  · simp only [hz, divisor_apply, DivisorOn.coe_neg, Pi.neg_apply]
+    rw [(hf z hz).order_inv]
+    by_cases ha : (hf z hz).inv.order = ⊤
+    · simp only [ha, WithTop.untopD_top, LinearOrderedAddCommGroupWithTop.neg_top, neg_zero]
+    lift (hf z hz).inv.order to ℤ using ha with a ha
+    rw [(by rfl : -a = (↑(-a) : WithTop ℤ)), WithTop.untopD_coe, WithTop.untopD_coe]
+    simp
```

Comment:
This line puzzles me slightly, since it stops working if you remove the `WithTop.untopD_coe` from the `rw` line, even though `WithTop.untopD_coe` is a simp lemma so it should be getting applied automatically by the `simp` in the next line anyway. It also doesn't work if you change `rw` to `simp_rw`, which is strange.

Reply:
I understand your question and I have no answer. Maybe @jcommelin can explain what's going on?

Reply:
I don't know exactly what is going on. But this line also closes the goal.
```suggestion
    simp [eq_comm (a := a), neg_eq_iff_eq_neg, WithTop.untopD_eq_iff]
```